### PR TITLE
upgrade metrics to support kafka 0.9

### PR DESCRIPTION
**Pull request by parasitew**
Add support for new producer and consumer in kafka 0.9.0 and it also supports kafka 0.8.2. The new producer/consumer will use NewStatsdMetricsReporter and NewStatsDReporter. New metrics are used for the producer.

to: @jun-he

### CODE REVIEW

**Comment by alexism** [comment]
- does the new release supports both 0.9+ and 0.8?
  - I think it is possible with minimum work.
- we need the latest exhaustive list of metrics (name, tags, meaning, etc). is this enough http://docs.confluent.io/1.0/kafka/monitoring.html ?
- Migration:
  - how to upgrade?
  - how to verify?
  - let’s highlight the expected changes (metric names, list, etc)
- it would be good to see a dump of the actual UDP data

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  private Map<String, KafkaMetric> kafkaMetrics;
+
+  private AbstractPollingReporter underlying = null;
+
+  public boolean isRunning() {
+    return running.get();
+  }
+
+  @Override
+  public void init(List<KafkaMetric> metrics) {
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: The initial list of metrics is not used.

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        }
+      };
+
+      registry.newGauge(NewStatsdMetricsReporter.class, name, strBuilder.toString() + "\n", gauge);
+      log.info("          metrics name: {}", name);
+    }
+  }
+
+  @Override
+  public void metricRemoval(KafkaMetric metric) {
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: - the method should not be flagged as @override
- the method is not used

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    } else {
+      log.warn("Reporter is disabled");
+    }
+  }
+
+  @Override
+  public void metricChange(final KafkaMetric metric) {
+    // Add new metrics in registry.
+    org.apache.kafka.common.MetricName metricName = metric.metricName();
+    String name = "kafka." + metricName.group() + "." + metricName.name();
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: should the prefix `kafka.` really be added?

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    if (!kafkaMetrics.containsKey(name)) {
+      Gauge<Double> gauge = new Gauge<Double>() {
+        @Override
+        public Double value() {
+          return metric.value();
+        }
+      };
+
+      registry.newGauge(NewStatsdMetricsReporter.class, name, strBuilder.toString() + "\n", gauge);
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: - why a line ending in the name?
- why not building a simpler and more explicit MetricName instance? `new MetricName(...)`

**Comment by alexism** [comment]
A more general design question:

The reporter implementation is only adding Gauges to the `com.yammer.metrics.Metrics#defaultRegistry`.
So why use an instance of yammer Registry when all is needed is a scheduled task that send Gauges?
see `com.airbnb.metrics.NewStatsDReporter#processGauge`

**Comment by alexism** [comment]
@parasitew ping

**Comment by parasitew** [comment]
Sorry for the late reply..

For the first question:
- This metrics reporter supports both 0.9 and 0.8 and I will put this in the README.
- I only get the metrics from new producer, will add other metrics from http://docs.confluent.io/1.0/kafka/monitoring.html in this PR
- Will update migration detail and new metrics output in the README

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 1 above)
Comment (parasitew) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: The 4 methods are declare in `MetricsReporter` ,  we have to override all of them but only need to get the updated metrics in `metricChange()` and send it to stats. https://github.com/apache/kafka/blob/6eacc0de303e4d29e083b89c1f53615c1dfa291e/clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java

``` java
  public void init(List<KafkaMetric> metrics);
  public void metricChange(KafkaMetric metric);
  public void metricRemoval(KafkaMetric metric);
  public void close();
```

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 3 above)
Comment (parasitew) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: The metrics look like this  `kafka.producer-topic-metrics.byte-rate:74.927084|g|#client-id=producer-1,topic=test`, I think it's easier for people to get all kafka-related metrics in datadog

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 2 above)
Comment (parasitew) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: Have to override it from `MetricsReporter` https://github.com/apache/kafka/blob/6eacc0de303e4d29e083b89c1f53615c1dfa291e/clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java

**Comment by parasitew** [comment]
The `defaultRegistry` is only for aggregating metrics. We put metrics to registry in `metricsChange()`, and every `epoch` time, statsd get all metrics from registry.

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 4 above)
Comment (parasitew) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: Just to make each metrics in udp output  in a line during debugging, should remove this.

**Comment by grindthemall** [comment]
I've been searching for something that does statsd gathering with Kafka 0.9.

I'm not enough of a developer to understand the state of this pull request, but if I'm reading it correctly, once merged will this support Kafka .9 stats?

**Comment by jun-he** [comment]
@grindthemall  Yes, currently, we are working on this PR and will ship it soon.

File: README.md [hunk 5]
```diff
@@ -17,35 +17,89 @@ This project provides a simple integration between Kafka and a StatsD reporter f
...
+- For Kafka `0.9.0.0` or later use `kafka-statsd-metrics2-0.5.0`
 - For Kafka `0.8.2.0` or later use `kafka-statsd-metrics2-0.4.0`
 - For Kafka `0.8.1.1` or prior use `kafka-statsd-metrics2-0.3.0`
 
 
 ## Releases
- 
+### 0.5.0
+
+ - `0.5.0` adds support for new producer/consumer in Kafka 0.9
```
Comment (jun-he) [inline README.md]: `support to report new producer/consumer metrics`?

File: README.md [hunk 6]
```diff
@@ -17,35 +17,89 @@ This project provides a simple integration between Kafka and a StatsD reporter f
...
+### 0.3.0
 
 - initial release
-        
+
 ## How to install?
 
 - [Download](https://bintray.com/airbnb/jars/kafka-statsd-metrics2/view) or build the shadow jar for `kafka-statsd-metrics`.
-- Install the jar in Kafka classpath, typically `./kafka_2.9.2-0.8.2.1/libs/`
+- Install the jar in Kafka classpath, typically `./kafka_2.9.2-0.9.0.1/libs/`
```
Comment (jun-he) [inline README.md]: update it to use the recommended version: `kafka_2.11-0.9.0.1`.

File: build.gradle [hunk 7]
```diff
@@ -23,7 +23,7 @@ repositories {
 
 dependencies {
     compile 'com.indeed:java-dogstatsd-client:2.0.11'
-    compile 'org.apache.kafka:kafka_2.10:0.8.2.1'
+    compile 'org.apache.kafka:kafka_2.10:0.9.0.1'
```
Comment (jun-he) [inline build.gradle]: use `kafka_2.11-0.9.0.1`

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+  public boolean isRunning() {
+    return running.get();
+  }
+
+  @Override
+  public void init(List<KafkaMetric> metrics) {
+    kafkaMetrics = new HashMap<String, KafkaMetric>();
+
+    if (enabled) {
+      startReporter(10);
```
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: Let's extract `10` and `kafka.` prefix etc. to constants.

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+import com.yammer.metrics.Metrics;
+import com.yammer.metrics.core.Gauge;
+import com.yammer.metrics.core.MetricName;
+import com.yammer.metrics.core.MetricsRegistry;
+import com.yammer.metrics.reporting.AbstractPollingReporter;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.slf4j.LoggerFactory;
+
+public class NewStatsdMetricsReporter implements MetricsReporter {
```
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: `New` is ambiguous.  Better to use an informative name.

File: src/main/java/com/airbnb/metrics/NewStatsDReporter.java [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+import static com.airbnb.metrics.Dimension.p95;
+import static com.airbnb.metrics.Dimension.p98;
+import static com.airbnb.metrics.Dimension.p99;
+import static com.airbnb.metrics.Dimension.p999;
+import static com.airbnb.metrics.Dimension.rate15m;
+import static com.airbnb.metrics.Dimension.rate1m;
+import static com.airbnb.metrics.Dimension.rate5m;
+import static com.airbnb.metrics.Dimension.stddev;
+
+public class NewStatsDReporter extends AbstractPollingReporter implements MetricProcessor<Long> {
```
Comment (jun-he) [inline src/main/java/com/airbnb/metrics/NewStatsDReporter.java]: Same here, `New` is unclear.

**Comment by jun-he** [comment]
@parasitew Can you add unit tests for all new features you added?  Thanks.

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 9 above)
Comment (alexism) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: agree with Jun

**Comment by alexism** [comment]
@parasitew regarding https://github.com/airbnb/kafka-statsd-metrics2/pull/16#issuecomment-215001953
I get that, but I think you don't need that registry. Simply schedule a thread that will send the metric values to statsd as gauges.

**Comment by jun-he** [comment]
@parasitew Per our discussion, I think it is a good idea to remove any dependency if possible.

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 9 above)
Comment (parasitew) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: Kafka use `NewShinyProducer/Consumer` 😓 , do you have a better name for this? https://github.com/apache/kafka/blob/6eacc0de303e4d29e083b89c1f53615c1dfa291e/core/src/main/scala/kafka/consumer/BaseConsumer.scala#L48

**Comment by parasitew** [comment]
@alexism 
Talked with Jun yesterday, we can use the  `defaultRegistry` for now because the old reporter still uses it. I will replace `com.yammer.metrics.Metrics#defaultRegistry` with a new thread/registry when we eliminate the old reporter.

**Comment by alexism** [comment]
@parasitew what "new registry" would be needed? you already maintain the list of metrics in `kafkaMetrics = new HashMap<String, KafkaMetric>();` The scheduled task simply has to iterate through that and send a statsd gauge.

can you explain what you get from the `Metrics#defaultRegistry` ?

File: src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java (hunk 1 above)
Comment (alexism) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: @parasitew this is not my understanding. 
- Based on the [documentation](https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/metrics/MetricsReporter.java#L25), "This is called when the reporter is first registered to _initially register_ all existing metrics". 
- This is confirmed by the `JmxReporter` implementation: https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java#L68-L69

The current implementation does not register anything on initialization, and any existing metrics will not be reported, only metrics created **after** the reporter initialization will be reported.

I think the confusion might come from the poor method name `metricChange`. The intent of that method is not to notify a change of the passed metric, but the creation of the passed metric.  `metricCreation` would have been more appropriate.

See also https://github.com/apache/kafka/blob/0.9.0/clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java#L249-L289

Comment (parasitew) [inline src/main/java/com/airbnb/kafka/NewStatsdMetricsReporter.java]: Thanks for pointing it out, I have fixed it.  The `metricChange` is quite confusing..

**Comment by parasitew** [comment]
@alexism Actually I just used the `AbstractPollingReporter` and it requires a `defaultRegistry`, I replaced `defaultRegistry` and `AbstractPollingReporter` with a `ScheduledExecutorService` to do the polling now
https://github.com/parasitew/kafka-statsd-metrics2/edit/tong-wei/update-metrics-to-support-0.9/src/main/java/com/airbnb/metrics/NewStatsDReporter.java?pr=/airbnb/kafka-statsd-metrics2/pull/16

btw, do you know why the build fails with openjdk6 and openjdk7 😢

**Comment by ismailfaruqi** [comment]
@parasitew so I could build this using Java 8 based on your branch and see the metrics reported to statsd? Btw I am using Kafka 0.9

**Comment by xuzha** [comment]
build failed probably because of this https://github.com/travis-ci/travis-ci/issues/5227#issuecomment-165131913

**Comment by xuzha** [comment]
@alexism I just pushed another commit. Please take a look, thanks so much.

**Comment by alexism** [comment]
LGTM thanks!

**Pull request merged**